File: jelly-vs-tower/typescript/src/Jelly.ts
import { ColorType } from './ColorType.js';
import { InvalidHealthException } from './InvalidHealthException.js';

export class Jelly {
  readonly id: string;
  readonly color: ColorType;
  private _health: number;

  constructor(id: string, color: ColorType, health: number) {
    if (health <= 0) throw new InvalidHealthException(health);
    this.id = id;
    this.color = color;
    this._health = health;
  }

  get health(): number {
    return this._health;
  }

  get isAlive(): boolean {
    return this._health > 0;
  }

  takeDamage(damage: number): void {
    this._health -= damage;
    if (this._health < 0) this._health = 0;
  }
}

File: jelly-vs-tower/csharp/src/JellyVsTower/Jelly.cs
namespace JellyVsTower;

public class Jelly
{
    public string Id { get; }
    public ColorType Color { get; }
    public int Health { get; private set; }
    public bool IsAlive => Health > 0;

    public Jelly(string id, ColorType color, int health)
    {
        if (health <= 0) throw new InvalidHealthException(health);
        Id = id;
        Color = color;
        Health = health;
    }

    public void TakeDamage(int damage)
    {
        Health -= damage;
        if (Health < 0) Health = 0;
    }
}
